File: day1-2.ml
open Std;;

let process str cur =
    match String.get str 0 with
    | '+' -> cur + int_of_string (String.sub str 1 ((String.length str) - 1))
    | '-' -> cur - int_of_string (String.sub str 1 ((String.length str) - 1))
    | _ -> cur
    
let rec calc strlist cur hsh =
   match strlist with
   | [] -> (false, cur)
   | h :: t -> 
      let out = process h cur in
        match Hashtbl.mem hsh out with
        | true -> (true, out)
        | false -> let _ = Hashtbl.add hsh out true in 
          calc t out hsh

let read_file filename =
  let chan = open_in filename in
  Std.input_list chan;;

exception Found_repeat;;

let p = 
  let inp = read_file "input1-1" in
    let hsh = Hashtbl.create 100000 in
      let _ = Hashtbl.add hsh 0 true in
        let cur = 0 in
          let curref = ref cur in  
            for i = 0 to 1000 do
              match calc inp !curref hsh with
              | (false, c) -> curref := c
              | (true, c) -> 
                let _ = print_endline (string_of_int c) in
                raise Found_repeat
            done;

File: day1-2.py
import sys

def main(args):
    lines = []
    with open("input1-1") as f:
        lines = f.readlines()

    hsh = {}
    cur = 0
    itr = 0

    hsh[cur] = True

    found = False
    while not found:
        itr += 1
        for l in lines:
            if l[0] == "+":
                cur += int(l[1:len(l)-1])
            else:
                cur -= int(l[1:len(l)-1])

            if cur in hsh:
                print(cur)
                found = True
                break

            hsh[cur] = True

    print("took iterations:", itr)

if __name__ == "__main__":
    main(sys.argv)
